### ISSUE TITLE
prometheus:stop() hangs when run from an ejabberd module

### ISSUE THREAD

**Issue by marcphilipp** [prometheus-erl/prometheus.erl#66]
First of all, thanks for creating this library, very useful!

I'm trying to implement an ejabberd module that uses `prometheus`:

Here's what I have:

```erlang
-module(mod_prometheus).

-include("ejabberd_http.hrl").
-include("logger.hrl").

-behavior(gen_mod).

-export([
  start/2,
  stop/1,
  depends/2,
  mod_opt_type/1,
  process/2
]).

start(Host, _Opts) ->
  ?INFO_MSG("~s starting on ~p", [?MODULE, Host]),
  prometheus:start().

stop(Host) ->
  ?INFO_MSG("~s stopping on ~p", [?MODULE, Host]),
  prometheus:stop(),
  ?INFO_MSG("~s stopped on ~p", [?MODULE, Host]),
  ok.

depends(_Host, _Opts) ->
  [].

mod_opt_type(_) ->
  [].

process([], #request{method = 'GET'}) ->
  {
    200,
    [{<<"Content-Type">>, prometheus_text_format:content_type()}],
    prometheus_text_format:format()
  }.
```

However, when I shutdown/restart ejabberd, `prometheus:stop()` never returns:

```
[...]
2017-05-24 17:55:34.446 [info] <0.63.0>@mod_prometheus:stop:30 mod_prometheus stopping on <<"localhost">>
```

Do you have an idea how I can find out what's going on? Any help would be highly appreciated!

**Comment by deadtrickster** [comment]
Hi, glad you found this lib useful :-) Will look into this shortly.

**Comment by marcphilipp** [comment]
I've done some further investigation. It looks like there's a deadlock because the `application_controller` is already busy stopping the ejabberd application and `handle_call` on it blocks indefinitely.

I guess I could just refrain from stopping the prometheus application completely. ejabberd does this as well for a bunch of internally used applications. But that doesn't really feel clean.

I've also tried to call `prometheus_sup:start_link()` directly, but then all the reports are empty. So I guess it has to be used as an application.

**Comment by deadtrickster** [comment]
Frankly I'm not that familiar with ejabberd plugin architecture (although I am (was) planning to write a plugin). So I need some time to learn it. The problem as I see from your latest comment is that this `stop` callback is called when main ejabberd app is stopping -> deadlock. Am I right?

Also I don't know what's your use-case, but there is https://github.com/deadtrickster/prometheus-httpd with `prometheus_http_impl` module just for implementing /metric endpoints.

**Comment by marcphilipp** [comment]
> Frankly I'm not that familiar with ejabberd plugin architecture (although I am (was) planning to write a plugin).

Interesting! Are you planning on collecting additional application-specific metrics like you did for RabbitMQ (e.g. online users per vhost)?

> The problem as I see from your latest comment is that this `stop` callback is called when main ejabberd app is stopping -> deadlock. Am I right?

Yep, I think so.

> Also I don't know what's your use-case, but there is https://github.com/deadtrickster/prometheus-httpd with prometheus_http_impl module just for implementing /metric endpoints.

Yeah, I've seen that. However, ejabberd has a built-in HTTP server that we're already using, so I want to expose it there. As you can see, it's not a lot of code (thanks to you).

**Comment by deadtrickster** [comment]
> Are you planning on collecting additional application-specific metrics like you did for RabbitMQ (e.g. online users per vhost)?

You know, it'd be really cool if you find time to write the plugin. I can help of course. 

> However, ejabberd has a built-in HTTP server that we're already using, so I want to expose it there

Yeah, but you missing content negotiation, compression and authentication. `prometheus_http_impl` module abstracts this and it's server-agnostic (e.g. https://gitlab.com/barrel-db/barrel-platform/blob/master/apps/barrel_monitor/src/barrel_monitor_exporter_handler.erl).

> Yep, I think so.

Can we somehow inject prometheus to ejabber supervision tree? You tried it and it didn't work?

**Comment by marcphilipp** [comment]
> You know, it'd be really cool if you find time to write the plugin. I can help of course.

Let me get a barebones plugin working locally and then we can talk about it. 🙂 

> Yeah, but you missing content negotiation, compression and authentication. `prometheus_http_impl` module abstracts this and it's server-agnostic.

I wasn't aware of `prometheus_http_impl`, thanks for the pointer!

> Can we somehow inject prometheus to ejabber supervision tree? You tried it and it didn't work?

I didn't add `prometheus_sup` to a supervisor, just linked it to the ejabberd process that starts modules and stopped `prometheus_sup` when the module is stopped. That produced empty output of `prometheus_text_format:format()`.

**Comment by deadtrickster** [comment]
Hi, how's it going? Maybe spawn_link a new process in stop/1, trap_exit and after 'main' process exits call prometheus:stop?

**Comment by marcphilipp** [comment]
ejabberd modules have to implement the `gen_mod` behavior which includes start/2 and stop/1. Right now, I don't stop it at all... which has no immediate implications but still feels wrong.

Would you mind sketching out your idea in a bit more detail?

**Comment by deadtrickster** [comment]
> Would you mind sketching out your idea in a bit more detail?
Yes, hopefully soon.

So stop/1 called with Host. Will it be called for all Hosts or just for one?

**Comment by marcphilipp** [comment]
It depends on how you register it. Since the general metrics are not host-specific, I worked around this by just registering it for a single host, i.e. `localhost`.

**Comment by deadtrickster** [comment]
Thanks, But can we stop ejabberd plugin for specific Host? or they stopped globally? In former case stopping prometheus isn't good for other running Hosts...

**Comment by marcphilipp** [comment]
Yes, you can stop plugins for a specific host. For me, that's not a problem. But for the general case that's undesirable. I guess it would be better to "attach" it to the ejabberd application but I don't think ejabberd provides a hook to do so.

**Comment by jpds** [comment]
I spent today working on a gatherer for ejabberd, and actually didn't get much further than the code in the first comment on my own.

There is also a basic exporter at https://github.com/skythet/ejabberd-prometheus-exporter/blob/master/mod_prometheus.erl - however I'd much rather use this library.

**Comment by NelsonVides** [comment]
Hey there,

Looking into this, the way to write a gatherer (or collector, using prometheus terminology), is to write a module implementing the `prometheus_collector` behaviour and, very importantly, register it in the sys.config, as in
```erl
    {prometheus, [{collectors, [default, your_custom_prometheus_collector]}}]},
```
or by using in your code
```erl
    ok = prometheus_registry:register_collector(your_custom_prometheus_collector).
```

Once registered, every scrape will use it to gather more data. You can see an example for RabbitMQ here: https://github.com/rabbitmq/rabbitmq-server/blob/expose-ra-counters/deps/rabbitmq_prometheus/src/collectors/prometheus_rabbitmq_raft_metrics_collector.erl#L31-L46, all it needs to do is end up returning a list of metrics created with `prometheus_model_helper:create_mf/4`. When stopping, you can use `prometheus_registry:deregister_collector/1`.

The prometheus application needs to have been started before all of this, the best way is to have added it as a dependency in your relx files, consult rebar3 or the usual BEAM packaging tooling.

Closing as it has been inactive for very long, feel free to reopen if there's still any doubt.
